src/context.spec.ts

```diff
@@ -116,14 +116,14 @@ describe('#getData()', () => {
     expect(context.getData({ requiredOnly: true })).toEqual([data[1]]);
   });
 
-  it('filters out undefineds and nulls when context optional with nullable = true', () => {
+  it('filters out undefineds and keeps nulls when context optional with nullable = true', () => {
     data[0].value = null;
     data[1].value = undefined;
 
     context = new ContextBuilder().setOptional({ checkFalsy: false, nullable: true }).build();
     context.addFieldInstances(data);
 
-    expect(context.getData({ requiredOnly: true })).toEqual([]);
+    expect(context.getData({ requiredOnly: true })).toEqual([data[0]]);
```

Comment:
Because right now `requiredOnly` actually does return data when the value is say an `Int`. So by your logic - that should actually not be returned as the Int was marked as optional and you passed requiredOnly. So I could argue the current logic is already breaking that contract. What my change is does is allow `null`s to come through when you specifically say you want to allow nulls ie: `nullable: true`

Reply:
Not quite. Perhaps the APIs are not very good here, but this behaviour shouldn't change IMO.

The value it has must match the definition of optional. That is, one of
- `undefined` (default behaviour)
- `null` or `undefined` (`nullable: true`)
- falsy (`checkFalsy: true` -- I don't like this name much tbh, it's express-validator v2 inheritance)

If your integer is optional but has value of 1, then it's considered as passed in the request.

Reply:
@gustavohenke 

I have the same problem.
I think it's more accurate to keep keys in both request body and validation schema in `matchedData` return value too.

But of course, it's breaking change.
How about add another parameter to matchedData function to keep optional keys which exist in request body?

Reply:
@gustavohenke 

https://github.com/express-validator/express-validator/issues/1273

I added use case here.
It's causing problem for partial update.